### PR TITLE
Feature/maps

### PR DESCRIPTION
**Pull request by cableman**
First version of maps.

It's not complete but we need to get it in to allow for styling.

### CODE REVIEW

File: web/modules/custom/genlyd_maps/images/marker.svg [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+<!-- Generator: Adobe Illustrator 18.0.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
+<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
+<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
+	 viewBox="0 0 365 560" enable-background="new 0 0 365 560" xml:space="preserve">
+<g>
+	<path fill="#00AEEF" d="M182.9,551.7c0,0.1,0.2,0.3,0.2,0.3S358.3,283,358.3,194.6c0-130.1-88.8-186.7-175.4-186.9
+		C96.3,7.9,7.5,64.5,7.5,194.6c0,88.4,175.3,357.4,175.3,357.4S182.9,551.7,182.9,551.7z M122.2,187.2c0-33.6,27.2-60.8,60.8-60.8
+		c33.6,0,60.8,27.2,60.8,60.8S216.5,248,182.9,248C149.4,248,122.2,220.8,122.2,187.2z"/>
+</g>
+</svg>
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/images/marker.svg]: Are both marker.svg and marker.png used?

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    positioning: 'top-center'
+  });
+  map.addOverlay(popup);
+
+  // Change mouse cursor when over a marker.
+  var target =  document.getElementById(map.getTarget());
+  map.on('pointermove', function(e) {
+    var pixel = map.getEventPixel(e.originalEvent);
+    var hit = map.hasFeatureAtPixel(pixel);
+    if (hit) {
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/js/genlyd_maps.js]: Why not move the map.hasFeatureAtPixel(pixel) into the if-statement?

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    else {
+      target.style.cursor = '';
+    }
+  });
+
+  // Display popup on click.
+  map.on('click', function(evt) {
+    // Loop over the features at the current event point.
+    map.forEachFeatureAtPixel(evt.pixel, function(feature, layer) {
+
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/js/genlyd_maps.js]: remove empty line?

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+      popup.setPosition(coordinates);
+
+      // Get properties and render them with twig into popup content.
+      var properties = feature.getProperties();
+      element.innerHTML = Twig.twig({ ref: 'popup' }).render(properties);
+
+      // Get hold of the close button and when clicked hide it.
+      var close = document.getElementsByClassName('js-close-button');
+      if (close.length) {
+        close[0].addEventListener('click', function() {
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/js/genlyd_maps.js:99]: Can this result in multiple click listeners registered?

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+          element.innerHTML = '';
+        });
+      }
+
+      // This is used to move the popup/map into view if this partly outside the
+      // current view area.
+      var bs_element = document.getElementsByClassName('popup')[0];
+      var clicked_pixel = evt.pixel;
+      var mapSize = map.getSize();
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/js/genlyd_maps.js]: mixed camelCase and snake_case notation for variable names

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+          center_px[0] -= width_left;
+        }
+
+        view.animate({
+          center: map.getCoordinateFromPixel(center_px),
+          duration: 300
+        });
+      }
+
+
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/js/genlyd_maps.js]: Empty lines

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+      // Get popup width.
+      var popup_width = Math.max(
+          bs_element.scrollWidth,
+          bs_element.offsetWidth,
+          bs_element.clientWidth
+        ) + offset_width;
+
+      // Calculate if the popup is outside the view area.
+      var height_left = mapSize[1] - clicked_pixel[1] - popup_height;
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/js/genlyd_maps.js]: What is the logic here? Can you explain it?

Comment (tuj) [reply web/modules/custom/genlyd_maps/js/genlyd_maps.js]: Maybe height_left is a bit misleading name?

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+      var width_left = mapSize[0] - clicked_pixel[0] - popup_width;
+
+      // Get current view and map center.
+      var view = map.getView();
+      var center_px = map.getPixelFromCoordinate(view.getCenter());
+
+      // Check if we are outside map view.
+      if (height_left < 0 || width_left < 0) {
+        if (height_left < 0) {
+          center_px[1] -= height_left;
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/js/genlyd_maps.js]: What do you do here?

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+        }
+        if (width_left < 0) {
+          center_px[0] -= width_left;
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/js/genlyd_maps.js]: See above comment

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+    });
+
+    var dataSource = new ol.source.Vector({
+      features: format.readFeatures(data, {
+        dataProjection: 'EPSG:4326',
+        featureProjection: 'EPSG:3857'
+      })
+    });
+
+    // Find the marker to use or fallback to default.
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/js/genlyd_maps.js:174]: Where is the fallback to default?

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+ *   The OpenLayers map object.
+ */
+function addOSMMap(map) {
+  map.addLayer(new ol.layer.Tile({
+    source: new ol.source.OSM()
+  }));
+}
+
+var map = initOpenlayersMap();
+//mapDebugInfo(map);
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/js/genlyd_maps.js]: commented out line

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+ * @param {ol.Map} map
+ *   The OpenLayers map object.
+ */
+function addOSMMap(map) {
+  map.addLayer(new ol.layer.Tile({
+    source: new ol.source.OSM()
+  }));
+}
+
+var map = initOpenlayersMap();
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/js/genlyd_maps.js:209]: Maybe add a comment to these 5 lines explaining that you initialise the map?

Comment (tuj) [reply web/modules/custom/genlyd_maps/js/genlyd_maps.js:209]: Should this be wrapped in a Drupal behavior?

File: web/modules/custom/genlyd_maps/js/templates/popup.html.twig [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+      <p class="popup--key">{{ Drupal.t('Address') }}</p>
+      <p class="popup--value">{{ address }}</p>
+    </div>
+    <div class="popup--metadata">
+      <p class="popup--key">{{ Drupal.t('Zipcode') }}</p>
+      <p class="popup--value">{{ zipcode }}</p>
+    </div>
+    <div class="popup--link-wrapper">
+      <a href="{{ url }}" target="_blank" class="popup--link">
+        <div class="popup--link-text">{{  Drupal.t('Further information') }}</div>
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/js/templates/popup.html.twig]: One space too many before Drupal.t

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+    /**
+     * Unsuccessful in acquiring user location.
+     *
+     * @param err
+     */
+    function error(err) {
+      /**
+       * @TODO: Find better design solution. There are 3 types of error codes.
+       */
+      alert('Unable to get position. Only works over https.');
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/js/genlyd_maps.js:252]: Can this be some popup on the map instead?

File: web/modules/custom/genlyd_maps/js/genlyd_maps_tabs.js [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+      mapTab.hide();
+      locationBtn.hide();
+
+      btn.click(function (event) {
+        event.preventDefault();
+        event.stopPropagation();
+
+        if (!mapTab.is(':visible')) {
+          btn.text( Drupal.t('Show table'));
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/js/genlyd_maps_tabs.js]: Space before Drupal.t

File: web/modules/custom/genlyd_maps/src/Controller/ApiController.php [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+   * Load all activities.
+   *
+   * @return \Drupal\genlyd_maps\Render\GeoJsonResponse
+   *   The activities as GeoJSON encoded array.
+   */
+  public function activates() {
+    $config = \Drupal::getContainer()->get('genlyd_maps.config')->getAll();
+
+    /**
+     * @TODO: Filter based on date?
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/src/Controller/ApiController.php:32]: @TODO or delete?

File: web/modules/custom/genlyd_maps/src/Form/GenlydMapSettingsForm.php [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    drupal_set_message('Settings saved');
+
+    // Set the rest of the configuration values.
+    $this->getBaseConfig()->setMultiple(array(
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/src/Form/GenlydMapSettingsForm.php]: Can this no just be:

$this->getBaseConfig()->set('genlyd_maps_google_api_key', $form_state->getValue('genlyd_maps_google_api_key'));

File: web/modules/custom/genlyd_maps/src/Form/GenlydMapSettingsForm.php [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+      'genlyd_maps_google_api_key' => $form_state->getValue('genlyd_maps_google_api_key'),
+    ));
+
+    drupal_flush_all_caches();
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/src/Form/GenlydMapSettingsForm.php]: Do we have a better solution than clearing all caches?

File: web/modules/custom/genlyd_maps/src/Plugin/Block/GenlydMaps.php [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+          ],
+        ],
+      ],
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function blockForm($form, FormStateInterface $form_state) {
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/src/Plugin/Block/GenlydMaps.php]: Is this necessary, when all we do is call the parent?

File: web/modules/custom/genlyd_maps/src/State/GenlydMapsConfig.php [hunk 20]
```diff
@@ -0,0 +1,21 @@
+<?php
+/**
+ * @file
+ * Contains key/value storage for itk_footer base config.
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/src/State/GenlydMapsConfig.php]: itk_footer?

File: web/modules/custom/genlyd_maps/templates/genlyd-maps-map-block.html.twig [hunk 21]
```diff
@@ -0,0 +1,9 @@
+<div class="js-map-tab-map">
+  <div class="map" style="height: 400px;width: 100%">
```
Comment (tuj) [inline web/modules/custom/genlyd_maps/templates/genlyd-maps-map-block.html.twig:2]: Inline styles?

**Review by tuj — CHANGES_REQUESTED** [review]

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js (hunk 7 above)
Comment (cableman) [reply web/modules/custom/genlyd_maps/js/genlyd_maps.js]: left = remaining

File: web/modules/custom/genlyd_maps/images/marker.svg (hunk 1 above)
Comment (cableman) [reply web/modules/custom/genlyd_maps/images/marker.svg]: Removed

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js (hunk 2 above)
Comment (cableman) [reply web/modules/custom/genlyd_maps/js/genlyd_maps.js]: Done

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js (hunk 4 above)
Comment (cableman) [reply web/modules/custom/genlyd_maps/js/genlyd_maps.js:99]: Nope

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js (hunk 3 above)
Comment (cableman) [reply web/modules/custom/genlyd_maps/js/genlyd_maps.js]: Done

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js (hunk 5 above)
Comment (cableman) [reply web/modules/custom/genlyd_maps/js/genlyd_maps.js]: Done

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js (hunk 12 above)
Comment (cableman) [reply web/modules/custom/genlyd_maps/js/genlyd_maps.js:209]: No as it don't have to wait for a attachement

File: web/modules/custom/genlyd_maps/js/genlyd_maps.js (hunk 11 above)
Comment (cableman) [reply web/modules/custom/genlyd_maps/js/genlyd_maps.js]: Comment added

File: web/modules/custom/genlyd_maps/js/templates/popup.html.twig (hunk 13 above)
Comment (cableman) [reply web/modules/custom/genlyd_maps/js/templates/popup.html.twig]: Done

File: web/modules/custom/genlyd_maps/src/Controller/ApiController.php (hunk 16 above)
Comment (cableman) [reply web/modules/custom/genlyd_maps/src/Controller/ApiController.php:32]: This is a placeholder for now... the code is not done yet

File: web/modules/custom/genlyd_maps/js/genlyd_maps_tabs.js (hunk 15 above)
Comment (cableman) [reply web/modules/custom/genlyd_maps/js/genlyd_maps_tabs.js]: done

File: web/modules/custom/genlyd_maps/src/Form/GenlydMapSettingsForm.php (hunk 17 above)
Comment (cableman) [reply web/modules/custom/genlyd_maps/src/Form/GenlydMapSettingsForm.php]: Done

File: web/modules/custom/genlyd_maps/src/Form/GenlydMapSettingsForm.php (hunk 18 above)
Comment (cableman) [reply web/modules/custom/genlyd_maps/src/Form/GenlydMapSettingsForm.php]: Comment added

File: web/modules/custom/genlyd_maps/src/Plugin/Block/GenlydMaps.php (hunk 19 above)
Comment (cableman) [reply web/modules/custom/genlyd_maps/src/Plugin/Block/GenlydMaps.php]: Removed

File: web/modules/custom/genlyd_maps/src/State/GenlydMapsConfig.php (hunk 20 above)
Comment (cableman) [reply web/modules/custom/genlyd_maps/src/State/GenlydMapsConfig.php]: Done

File: web/modules/custom/genlyd_maps/templates/genlyd-maps-map-block.html.twig (hunk 21 above)
Comment (cableman) [reply web/modules/custom/genlyd_maps/templates/genlyd-maps-map-block.html.twig:2]: yes, as there is not design for this yet

**Comment by cableman** [comment]
Request for re-review

File: web/modules/custom/genlyd_maps/templates/genlyd-maps-map-block.html.twig (hunk 21 above)
Comment (tuj) [reply web/modules/custom/genlyd_maps/templates/genlyd-maps-map-block.html.twig:2]: ok

**Pull request merged**